File: 04-arrow/src/main/java/Main.java
import io.vavr.control.Try;
import java.io.FileInputStream;

public class Main {
    public static double divide(double a, double b) {
        return a / b;
    }

    public static Try<Double> divide2(Double a, Double b) {
        Double res = a / b;
        if(res == Double.POSITIVE_INFINITY) {
            return Try.failure(new Exception("😡 Infinity"));
        }
        else {
            return Try.success(res);
        }
    }

    public static Try<String> readFile(String filePath) {
        return Try.of(() -> {
            FileInputStream fis = new FileInputStream(filePath);
            byte[] b = new byte[100];
            fis.read(b, 0, 100);
            return new String(b);
        });
    }

    public static void main(String[] args) {
        System.out.println(divide(5.0, 0.0));

        divide2(5.0, 0.0)
        .onFailure(it -> System.out.println(it.getMessage()))
        .onSuccess(it -> System.out.println(it));

        readFile("test.txt")
        .onFailure(it -> System.out.println(it.getMessage()))
        .onSuccess(it -> System.out.println(it));
    }
}

File: 04-arrow/src/main/kotlin/main.kt
package demo

import arrow.core.Failure
import arrow.core.Success
import arrow.core.Try
import java.io.FileInputStream

fun divide(a: Double, b: Double): Double {
  return a/b
}

// Try retourne soit un Failure, soit un Success
fun divide2(a: Double, b: Double): Try<Double> {
  val res = a/b
  return when(res) {
    Double.POSITIVE_INFINITY -> Failure(Exception("😡 Infinity"))
    else -> Success<Double>(res)
  }
}

// autre façon de l'utiliser

fun readFile(filePath: String): Try<String> {
  val fis = Try { FileInputStream(filePath) }
  return fis.map { it.readBytes().toString() }
}

fun main(args: Array<String>) {

  println(divide(5.0, 0.0))

  divide2(5.0, 0.0).let {
    when(it) {
      is Failure -> println(it.exception.message)
      is Success -> println(it.value)
    }
  }

  readFile("test.txt").let {
    when(it) {
      is Failure -> println(it.exception.message)
      is Success -> println(it.value)
    }
  }
}